velox/expression/FunctionSignature.cpp

```diff
@@ -212,9 +212,54 @@ FunctionSignature::FunctionSignature(
       argumentTypes_{std::move(argumentTypes)},
       constantArguments_{std::move(constantArguments)},
       variableArity_{variableArity} {
-  validate(variables_, returnType_, argumentTypes_, constantArguments_);
+  validate(variables_, returnType_, argumentTypes_, constantArguments_, {});
 }
 
+FunctionSignature::FunctionSignature(
+    std::unordered_map<std::string, SignatureVariable> variables,
+    facebook::velox::exec::TypeSignature returnType,
+    std::vector<TypeSignature> argumentTypes,
+    std::vector<bool> constantArguments,
+    bool variableArity,
+    const std::function<std::unordered_set<std::string>(
+        const std::unordered_map<std::string, SignatureVariable>&
+            typeVariables)>& usedVariables)
+    : variables_{std::move(variables)},
+      returnType_{std::move(returnType)},
+      argumentTypes_{std::move(argumentTypes)},
+      constantArguments_{std::move(constantArguments)},
+      variableArity_{variableArity} {
+  validate(
+      variables_,
+      returnType_,
+      argumentTypes_,
+      constantArguments_,
+      usedVariables(variables_));
+}
+
+AggregateFunctionSignature::AggregateFunctionSignature(
+    std::unordered_map<std::string, SignatureVariable> variables,
+    facebook::velox::exec::TypeSignature returnType,
+    facebook::velox::exec::TypeSignature intermediateType,
+    std::vector<TypeSignature> argumentTypes,
+    std::vector<bool> constantArguments,
+    bool variableArity)
+    : FunctionSignature(
+          std::move(variables),
+          std::move(returnType),
+          std::move(argumentTypes),
+          std::move(constantArguments),
+          variableArity,
+          [&intermediateType](
+              const std::unordered_map<std::string, SignatureVariable>&
+                  typeVariables) {
```

Comment:
I'm not following this design. Would you clarify?

Reply:
@mbasmanova When validating a `FunctionSignature`, it will collect all variables used in the "Type"s and compare this `usedVariables` to the size of `variables_`. If an inconsistency is found, indicating that `variables_` contains unused variables, the `FunctionSignature` is considered invalid.

The `AggregateFunctionSignature` inherits from `FunctionSignature` and adds a new "type" `intermediateType_`. If i`ntermediateType_` uses variables, it will be stored in the `variables_` of `FunctionSignature`. However, during validation of `FunctionSignature`, variables used in `intermediateType_` are not collected. This leads to a situation where the size of the collected `usedVariables` during the `FunctionSignature` validation is less than `variables_.size`, resulting in it being deemed invalid.

Therefore, when validating an `AggregateFunctionSignature`, we need to introduce a lambda function. This lambda function will first validate `intermediateType_` and collect its used variables into `usedVariables`. This `usedVariables`, which now includes the variables from `intermediateType_`, will then be passed as an argument to the new constructor of `FunctionSignature` for validation. In this way, `FunctionSignature` receives a `usedVariables` that already contains the variables from `intermediateType_`. Consequently, when checking if `usedVariables.size` matches `variables_.size`, no issues will arise.

Reply:
@liujiayi771 Thank you for clarifying. 

> If the subclass introduces a new Type variable and that Type uses signature variables

If that's the only use case, then maybe a simple solution is to pass a list of additional types to the base constructor for validation. What do you think?

Reply:
@mbasmanova We need to use `validateBaseTypeAndCollectTypeParams` to collect the used variables in `intermediateType` first, and then pass it to the constructor of `FunctionSignature`. Can we achieve this without using lambda functions?

Reply:
@mbasmanova I understand what you mean. If it's about passing the type, I think this way is simpler.